subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareInputStreamHasher.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import org.apache.commons.io.input.CloseShieldInputStream;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.Hashing;
+import org.gradle.internal.hash.PrimitiveHasher;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Optional;
+
+public class LineEndingAwareInputStreamHasher {
+    private static final HashCode SIGNATURE = Hashing.signature(LineEndingAwareInputStreamHasher.class);
+    private static final int BUFFER_SIZE = 8192;
+
+    public Optional<HashCode> hashContent(InputStream inputStream) throws IOException {
+        return hash(inputStream);
+    }
+
+    public Optional<HashCode> hashContent(File file) throws IOException {
+        try (FileInputStream inputStream = new FileInputStream(file)) {
+            return hashContent(inputStream);
+        }
+    }
+
+    /**
+     * Returns empty if the file is detected to be a binary file
+     */
+    private Optional<HashCode> hash(InputStream inputStream) throws IOException {
```

Comment:
Would be nice to add some unit tests for this method. It is the core of the whole PR with some non-trivial logic. This could replace some of the duplicate tests in `LineEndingAwareResourceHasherTest` and `LineEndingAwareFileSystemLocationSnapshotHasherTest`.

Reply:
Done.